ansible/roles/setup_todo_app/tasks/main.yaml

```diff
@@ -0,0 +1,54 @@
+- name: Create directory for app
+  file:
+    name: /opt/todoapp
+    state: directory
+    owner: ec2-user
+    group: ec2-user
+    mode: '755'
+
+- name: Check if /opt/todoapp is empty
+  find:
+    paths: /opt/todoapp
+    file_type: directory
+  register: files
+
+- name: Download to-do app from github
+  git:
+    repo: 'https://github.com/MHolmes219/DevOps-Course-Starter.git'
+    dest: /opt/todoapp
+  when: not files.files
+
+- name: Update todoapp permissions
+  command:
+    cmd: chown -R ec2-user:ec2-user /opt/todoapp
+
+- name: Add gunicorn with poetry
+  shell:
+    cmd: poetry add gunicorn
+  args:
+    chdir: /opt/todoapp
+  become: false
+
+- name: Run poetry install
+  shell:
+    cmd: poetry install
+  args:
+    chdir: /opt/todoapp
+  become: false
+
+- name: Create .env file
+  template:
+    src: templates/.env.j2
+    dest: /opt/todoapp/.env
+    owner: ec2-user
+    group: ec2-user
+    mode: '644'
+
+- name: Create todo app service
+  template:
+    src: templates/todoapp.service.j2
+    dest: /etc/systemd/system/todoapp.service
+    mode: '644'
+    owner: root
+    group: root
+  notify: enable_todo_service
```

Comment:
Using notify for the service restart task is nice (to avoid unnecessary restarts). But I think you're missing some more notify tags. E.g. if the .env file or app code change, shouldn't we also restart the service?